aptos-move/mvhashmap/src/lib.rs

```diff
@@ -105,30 +164,88 @@ impl<K: Hash + Clone + Eq, V> MVHashMap<K, V> {
         map.remove(&txn_idx);
     }
 
-    /// read may return Ok((Arc<V>, txn_idx, incarnation)), Err(dep_txn_idx) for
-    /// a dependency of transaction dep_txn_idx or Err(None) when no prior entry is found.
-    pub fn read(&self, key: &K, txn_idx: TxnIndex) -> Result<(Version, Arc<V>), Option<TxnIndex>> {
+    /// Read entry from transaction 'txn_idx' at access path 'key'.
+    pub fn read(&self, key: &K, txn_idx: TxnIndex) -> Result<V> {
+        use MVHashMapError::*;
+        use MVHashMapOutput::*;
+
         match self.data.get(key) {
             Some(tree) => {
-                // Find the dependency
                 let mut iter = tree.range(0..txn_idx);
-                if let Some((idx, write_cell)) = iter.next_back() {
-                    let flag = write_cell.flag();
+
+                // If read encounters a delta, it must traverse the block of
+                // transactions (top-down) until it encounters a write or reaches
+                // the end of the block. During traversal, all deltas have to be
+                // aggregated together.
+                let mut aggregator: Option<DeltaOp> = None;
+                while let Some((idx, entry)) = iter.next_back() {
+                    let flag = entry.flag();
 
                     if flag == FLAG_ESTIMATE {
                         // Found a dependency.
-                        Err(Some(*idx))
+                        return Err(Dependency(*idx));
                     } else {
+                        // The entry should be populated.
                         debug_assert!(flag == FLAG_DONE);
-                        // The entry is populated, return its contents.
-                        let write_version = (*idx, write_cell.incarnation);
-                        Ok((write_version, write_cell.data.clone()))
+
+                        match &entry.cell {
```

Comment:
it's more idiomatic to write like this
```
  match (aggregator.as_ref(), &entry.cell) => {
    (Some(delta), EntryCell::Write(..) => ..,
    (None, EntryCell::Write(..) => ..,
    (Some(delta), EntryCell::Delta(..) => ..,
    ...
  }
```

Reply:
the logic changed somewhat where some conditions are shared and this might be a bit verbose.

Reply:
done